File: Sorting/Quick Sort/C/Quick_Sort.c
#include<stdio.h>
#include<stdlib.h>

void printarr(int);
void quicksort(int,int);
int partition(int,int);
void swap(int *, int *);

int *a;

int main()
{
    int i,n;

    printf("Enter the size of the array:\n");
    scanf("%d",&n);
    a=(int*) calloc(n,sizeof(int));

    printf("Enter the array:\n");
    for (i=0; i<n; i++)
    {
        printf("Element %d: ",i+1);
        scanf("%d",&a[i]);
    }

    printf("Before sorting:\n");
    printarr(n);

    quicksort(0,n-1);

    printf("After sorting:\n");
    printarr(n);

    return 0;
}

void quicksort(int beg, int end)
{
    int pivot;
    if(beg<end)
    {
        pivot=partition(beg,end);
        quicksort(beg,pivot-1);
        quicksort(pivot+1,end);
    }
}

int partition(int beg, int end)
{
    int l,r,pivot,flag=0;
    l=pivot=beg;
    r=end;

    while(flag==0)
    {
        while((pivot!=r)&&(r>=l))
        {
            if(a[r]<=a[pivot])
                {
                    swap(&a[r],&a[pivot]);
                    pivot=r;
                    break;
                }
            r--;
        }
        if(r==l)
            flag=1;

        while((l!=pivot)&&(l<=r)&&(flag!=1))
        {
            if(a[l]>=a[pivot])
            {
                swap(&a[l],&a[pivot]);
                pivot=l;
                break;
            }
            l++;
        }
        if(l==r)
            flag=1;
    }

    return pivot;
}

void swap(int *a, int *b)
{
    *a=*a+*b;
    *b=*a-*b;
    *a=*a-*b;
}

void printarr(int n)
{
    int i;
    for (i=0; i<n; i++)
    {
        printf("Element %d: %d\n",i+1,a[i]);
    }
}

File: Sorting/Quick Sort/Python/QuickSort.py
#Implementation of QuickSort in Python
def partition(arr,low,high):
    # index of smaller element
    i = (low-1)

    # Pivot is the last element in arr
    pivot = arr[high]
    print("Pivot is: " + str(pivot))

    for j in range(low,high):
        # Checks if the current element is smaller than or
        # equal to pivot
        if arr[j] <= pivot:

            # Increment index of smaller element
            i +=1

            # Swap the values at the high index and the smaller index
            arr[i],arr[j] = arr[j],arr[i]

    # Swap the values
    arr[i+1],arr[high] = arr[high],arr[i+1]
    return (i+1)

# arr is the array to be sorted
# low is the starting index
# high is the ending index
def QuickSort(arr,low,high):
    if len(arr) == 1:
        return arr
    if low < high:

        # p is partitioning index, arr[p] is at the right place
        p = partition(arr,low,high)

        # Sperately sort elements before partition and after partition
        QuickSort(arr,low,p-1)
        QuickSort(arr,p+1,high)

if __name__=='__main__':
    arr = [10,7,8,9,1,5]
    n = len(arr)
    QuickSort(arr,0,n-1)
    print("Sorted array is:")
    for i in range(n):
        print(arr[i])